transports/webtransport-websys/src/fused_js_promise.rs

```diff
@@ -0,0 +1,59 @@
+use futures::FutureExt;
+use js_sys::Promise;
+use std::task::{ready, Context, Poll};
+use wasm_bindgen::JsValue;
+use wasm_bindgen_futures::JsFuture;
+
+/// Convenient wrapper to poll a promise to completion.
+#[derive(Debug)]
+pub(crate) struct FusedJsPromise {
+    promise: Option<JsFuture>,
+}
+
+impl FusedJsPromise {
+    /// Creates new uninitialized promise.
+    pub(crate) fn new() -> Self {
+        FusedJsPromise { promise: None }
+    }
+
+    /// Initialize promise if needed and then poll.
+    ///
+    /// If promise is not initialized then `init` is called to initialize it.
+    pub(crate) fn maybe_init_and_poll<F>(
+        &mut self,
+        cx: &mut Context,
+        init: F,
+    ) -> Poll<Result<JsValue, JsValue>>
+    where
+        F: FnOnce() -> Promise,
+    {
+        if self.promise.is_none() {
+            self.promise = Some(JsFuture::from(init()));
+        }
+
+        self.poll(cx)
+    }
```

Comment:
How about we change this to just take the callback and return a mutable reference to itself? I think that composes slightly better because we can just chain `.poll` onto it.

Reply:
To be used like this?

```rust
            ready!(self
                .writer_ready_promise
                .maybe_init(|| self.writer.ready())
                .poll(cx))
```

Reply:
Yep!

Reply:
Sound good! I did the change